Adding an example for Supervised Contrastive Learning

ksalama opened the pull request:
Supervised Contrastive Learning: https://arxiv.org/abs/2004.11362

ksalama commented:
@fchollet - kindly review the PR

fchollet reviewed (commented):
Thank you for the PR. Please only include the Python file in the PR for now.

ksalama commented:
Done - Only the Python file is included in the PR.

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,204 @@
+
+
+def create_encoder():
+    return tf.keras.Sequential(
```
Prefer using the functional API everywhere

ksalama replied:
Done

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,204 @@
+encoder = create_encoder()
+encoder.summary()
+
+BATCH_SIZE = 256
```
The code mixes lowercase global constants and uppercase global constants. You can use lowercase everywhere.

ksalama replied:
Done

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,204 @@
+            tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
+            tf.keras.layers.Conv2D(64, kernel_size=(3, 3), activation="relu"),
+            tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
+            tf.keras.layers.Flatten(),
```
Prefer using a GlobalAveragePooling layer here.

This architecture is overall too small for CIFAR10. Prefer using at least once more conv stage and more filters per conv layer. Also use 2 conv layers per stage (so more conv than maxpool)

ksalama replied:
Done - Thanks for the suggestion. It improved the results big time, and showed how the Supervised Contrastive Learning can be very powerful!

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,204 @@
+    y=y_train,
+    batch_size=BATCH_SIZE,
+    epochs=NUM_EPOCHS,
+    validation_data=(x_test, y_test),
```
Prefer using a different dataset for validation than for the final test, so here you can use `validation_split=0.1` for instance

ksalama replied:
done

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,204 @@
+    y=y_train,
+    batch_size=BATCH_SIZE,
+    epochs=NUM_EPOCHS,
+    validation_data=(x_test, y_test),
```
Likewise, use validation_split

ksalama replied:
Done

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,204 @@
+)
+
+accuracy = classifier.evaluate(x_test, y_test)[1]
+print(f"Test accuracy: {round(accuracy*100, 2)}%")
```
Add spaces around operators (everywhere)

ksalama replied:
Done

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,204 @@
+"""
+## Experiment 1: Train the baseline classification model
+
+In this experiment, a baseline classifier is trained normally, i.e., the encoder and the
```
For each experiment, please add a short text paragraph describing what the experiment does, and what it shows. Consider providing references to publications

ksalama replied:
I am referencing the publication in the introduction. I also added description to the second experiment as it was missing.

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,204 @@
+    for layer in encoder.layers:
+        layer.trainable = trainable
+
+    inputs = tf.keras.Input(shape=input_shape)
```
Prefer importing:

- The `layers` module for access to layers
- `keras` for everything else

ksalama replied:
Done

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,204 @@
+# the data, split between train and test sets
+(x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()
+
+y_train, y_test = tf.squeeze(y_train), tf.squeeze(y_test)
```
This line is unexplained

ksalama replied:
Explanation added.

fchollet replied:
This isn't really necessary, is it? What happens if you remove the line?

ksalama replied:
Seems like it isn't :-) 
Removed.

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,204 @@
+Essentially, training an image classification model with Supervised Contrastive Learning
+is peformed in two phases:
+
+  1. Pre-training an encoder to generate feature vectors for input images such that feature
```
Unindent the list so that it renders properly

ksalama replied:
Done

ksalama commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,204 @@
+"""
+
+
+def make_supervised_contrastive_loss_fn(temperature=1):
```
Done

ksalama commented:
@fchollet - Have addressed your comments. Kindly review my revisions.
Thanks!

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,186 @@
+Description: Using supervised contrastive learning for image classification.
+"""
+
+## Introduction
```
Missing a """ above

ksalama replied:
Done

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,186 @@
+
+## Introduction
+
+[Supervised Contrastive Learning](https://arxiv.org/abs/2004.11362) (Prannay Khosla et al.) is a training methodology that outperforms cross-entropy on supervised learning tasks. 
```
Please keep text short (helps with code reviews)

ksalama replied:
Could you suggest a shorter text for the intro?

fchollet replied:
Sorry, I meant to say "keep text lines short" (hence the comment about code review). Please shorten all lines to be no longer than ~80-90 characters.

ksalama replied:
done

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,186 @@
+
+[Supervised Contrastive Learning](https://arxiv.org/abs/2004.11362) (Prannay Khosla et al.) is a training methodology that outperforms cross-entropy on supervised learning tasks. 
+
+Essentially, training an image classification model with Supervised Contrastive Learning is peformed in two phases: 
```
"performed". Please proofread for typos

ksalama replied:
Done

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,186 @@
+## Setup
+"""
+
+!pip install tensorflow-addons
```
A shell line should be put in a shell block, and should not start with ! since it's added automatically

ksalama replied:
Done

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,186 @@
+import numpy as np
+import keras
+from keras import layers
+from keras.losses import Loss
```
Just use `keras` for this, don't import `Loss`

ksalama replied:
Done

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,186 @@
+  for layer in encoder.layers:
+    layer.trainable = trainable
+
+  inputs = tf.keras.Input(shape=input_shape)
```
keras

ksalama replied:
Done

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,186 @@
+
+  inputs = tf.keras.Input(shape=input_shape)
+  features = encoder(inputs)
+  features = tf.keras.layers.Dropout(dropout_rate)(features)
```
layers

ksalama replied:
Done

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,186 @@
+            tf.transpose(feature_vectors_normalized)),
+        temperature
+    )
+
```
Remove blank lines

ksalama replied:
Done

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,186 @@
+
+encoder = create_encoder()
+encoder.compile(
+    optimizer=tf.keras.optimizers.Adam(),
```
You can use `keras.optimizers`

ksalama replied:
Done

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,186 @@
+accuracy = classifier.evaluate(x_test, y_test)[1]
+print(f'Test accuracy: {round(accuracy*100, 2)}%')
+
+"""We get to ~69.8% validation accuracy.
```
I'm still concerned that this is a very low score. We need to improve the model. Maybe you could try using a ResNet like architecture. The baseline score should be at least 80-85%. You could also try using KerasTuner to find a good model configuration.

70% accuracy can be done without even using deep learning btw.

(also, this should be "test accuracy", not "validation accuracy")

ksalama replied:
Hi @fchollet - I think the purpose is to show the benefit of Supervised Constative Learning (SCL) over the conventional learning technique, rather than producing state of the art results for the cifar10. I think the model is more complex and produced better results in the conventional (baseline) model, it will produce even better results in the  SCL approach.

Also note that the model, in both cases, is only trained for 25 epochs.

If you agree with the rationale, I can add a sentence in the intro explaining this.

fchollet replied:
CIFAR10 state of the art is 99%. I am not suggesting you should aim for the state of the art. I suggest getting "decent" results, which for a deep learning model is around 80-85%.

70% is far too low because it can be achieved without a convnet.

fchollet replied:
@ksalama what do you think?

ksalama replied:
@fchollet Ok - I used data augmentation and keras.applications.ResNet50V2 (random wrights)
- with 50 epochs, the baseline model achieves ~76% test acc, compared to ~81% with the SCL
- with 100 epochs, the baseline model achieves ~80 test acc, compared to ~82% with the SCL

It seems that ResNet50V2 is a bit big for cifar10 and tends to overfit the training dataset. However, code-wise, it is clean to use it as the encoder. In addition, ResNet50 was the model used in the research paper. I saw that a 90% test acc can be reach with data augmentation and a right model architecture[1]. However, as I mentioned, I wanted to illustrate how easy it is to implement SCL, and how it boost the performance of the model even with small number of epochs. I think SCL pays-off with "harder" datasets that required more complex models.
 
What are your thoughts? Do you think these experiments are good enough? If not, I can try the following:
1- Implement the architecture in [1]
2- Use another dataset (maybe cifar10) in which SCL can payoff more.

[1] https://github.com/moritzhambach/Image-Augmentation-in-Keras-CIFAR-10-/blob/master/CNN%20with%20Image%20Augmentation%20(CIFAR10).ipynb

fchollet replied:
> It seems that ResNet50V2 is a bit big for cifar10 and tends to overfit the training dataset

Using this type of architecture is a good idea. You could simplify the model a bit (fewer blocks, smaller blocks) to right-size it for CIFAR10 (this architecture was developed for ImageNet which is much larger).

Here's a simple architecture you could use, a kind of mini-xception model: https://colab.research.google.com/drive/1qeFE_xkC1qnbbrJkmAw3hRFspu-d_OVn?usp=sharing

(you can tune the number of blocks and their size easily, e.g. with KerasTuner)

ksalama replied:
Thank you for sharing the mini-xception notebook! 

In this example, I think it would be nice to show how to use one of the Keras application models as-is to be the encoder, especially as the focus of the tutorial is not to show how to build a powerful vision architecture, rather how SCL can improve the performance of "a given" architecture. Thus, using an out-of-the-box model (like ResNet50V2) will make the focus of the tutorial on the SCL technique, rather than the details of the fancy architecture of mini-xception :-)

So my preference is use the current implementation for the tutorial. However, if you think it is necessary for tutorial to be published, I am happy to try this mini-xception model.

What do you think?

fchollet reviewed (commented):
Thank you for the update!

ksalama commented:
@fchollet kindly find my revisions. Cheers

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,206 @@
+
+"""## Using image data augmentation"""
+
+data_generator = ImageDataGenerator(
```
ImageDataGenerator is legacy, don't use it. Instead, use preprocessing layers:

- `Normalization`: https://keras.io/api/layers/preprocessing_layers/core_preprocessing_layers/normalization/
- data augmentation layers (see example https://keras.io/examples/vision/image_classification_from_scratch/#using-image-data-augmentation)

ksalama replied:
@fchollet - I tried using these APIs, but the performance (train and test) wasn't as good as using the ImageDataGenerator.

Basically I did the following:

```
data_augmentation = keras.Sequential(
    [
      layers.experimental.preprocessing.Normalization(),
      layers.experimental.preprocessing.RandomFlip("horizontal"),
      layers.experimental.preprocessing.RandomRotation(0.2),
      layers.experimental.preprocessing.RandomWidth(0.2),
      layers.experimental.preprocessing.RandomHeight(0.2)
    ]
)

def create_encoder():
  resnet = keras.applications.ResNet50V2(
      include_top=False,
      weights=None,
      input_shape=input_shape,
      pooling='avg')

  inputs = keras.Input(shape=input_shape)
  augmented = data_augmentation(inputs)
  outputs = resnet(augmented)
  model = keras.Model(inputs=inputs, outputs=outputs, name="cifar10-encoder")
  return model

encoder = create_encoder()
encoder.summary()

...

history = classifier.fit(
  x=x_train, 
  y=y_train, 
  batch_size=batch_size)
```

The test accuracy doesn't reach 70% in the baseline model. You can validate this using this [Colab](https://colab.research.google.com/drive/1z3EzFqIhZIIfnkQMulEIZRNJVq92eCYk#scrollTo=Fnt7P8C05mzs).

Do you have any suggestions around this?

fchollet replied:
There are several considerations:

- `Normalization` needs its state to be set. You can do this by calling `layer.adapt(x_train)` after creating the layer.
- The image data augmentation may not be the same. Make sure to visualize the output images and adjust the amount of random rotation / etc accordingly. The meaning of the arguments has changed between the 2 APIs.

ksalama replied:
Fixed. Thanks for the guidance!

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,206 @@
+
+  inputs = keras.Input(shape=input_shape)
+  outputs = densenet(inputs)
+  model = keras.Model(inputs=inputs, outputs=outputs, name="cifar10-encoder")
```
This wrapper model does not add anything, you can just use `ResNet50V2` as is.

fchollet replied:
(also densenet != ResNet50V2)

ksalama replied:
Fixed

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,206 @@
+  features = encoder(inputs)
+  features = layers.BatchNormalization()(features)
+  features = layers.Dropout(dropout_rate)(features)
+  features = layers.Dense(hidden_units, 'relu')(features)
```
You should not, in general, do relu before BN. You should do it after.

ksalama replied:
Done

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,206 @@
+  model.compile(
+    optimizer=keras.optimizers.Adam(learning_rate),
+    loss=keras.losses.SparseCategoricalCrossentropy(),
+    metrics=keras.metrics.SparseCategoricalAccuracy()
```
This argument should be a list.

ksalama replied:
Done

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,206 @@
+def add_projection_head(encoder):
+  inputs = keras.Input(shape=input_shape)
+  features = encoder(inputs)
+  outputs = layers.Dense(projection_units, 'relu')(features)
```
Same comment about relu

ksalama replied:
Done

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,206 @@
+accuracy = classifier.evaluate(data_generator.flow(x_test, y_test))[1]
+print(f'Test accuracy: {round(accuracy * 100, 2)}%')
+
+"""We get to ~81.3% test accuracy."""
```
Please add a comment about comparing the two methods, and maybe some pointers to which factors affect when to use SCL and when not to use it.

ksalama replied:
@fchollet - I added a 'conclusion' section at the end. LMK if this what you had in mind.

fchollet reviewed (commented):
Looking good! Just a few minor comments.

fchollet reviewed (commented):
Thank you! Looks good, please use preprocessing layers and address the remaining comments.

ksalama commented:
@fchollet - I think I have addressed all your comments, including:

1. Using preprocessing layers for data augmentation (maintaining the same or slightly better results compared to the DataGenerator)
2. Adding a conclusion section
3. Fixing the ReLU and BatchNorm layers ordering

Please let me know if there is anything else need to be modified.

ksalama commented:
I have simplified the models by removing the BatchNorm layers as they doesn't improve the results

fchollet commented:
> Using preprocessing layers for data augmentation (maintaining the same or slightly better results compared to the DataGenerator)

The code I'm seeing still uses ImageDataGenerator; did you push the changes to the PR?

ksalama commented:
@fchollet - Sorry for the confusion. I've just pushed the changes and they should be now part of the PR

ksalama commented:
Also made some formatting updates

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,234 @@
+  inputs = keras.Input(shape=input_shape)
+  features = encoder(inputs)
+  features = layers.Dropout(dropout_rate)(features)
+  features = layers.Dense(hidden_units, 'relu')(features)
```
pass the activation as a keyword arg, `activation='relu'` (easier to read)

ksalama replied:
Done

fchollet commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,234 @@
+def add_projection_head(encoder):
+  inputs = keras.Input(shape=input_shape)
+  features = encoder(inputs)
+  outputs = layers.Dense(projection_units, 'relu')(features)
```
Likewise here

ksalama replied:
Done

fchollet reviewed (approved):
Looks great! Please address the remaining nits and add the autogenerated `.md` and `.ipynb` files (you can generate them with the `python autogen.py add_example command`)

ksalama commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,234 @@
+import numpy as np
+import keras 
+from keras import layers
+from keras.preprocessing.image import ImageDataGenerator
```
Done

ksalama commented:
Remaining nits addressed

fchollet reviewed (approved):
Looks good, please add the generated files to the PR!

8bitmp3 commented on examples/vision/supervised-contrastive-learning.py:
```diff
@@ -0,0 +1,238 @@
+outperformed the conventional technique in terms of the test accuracy. Note that
+the same training budget (i.e., number of epochs) was given to each technique.
+Supervised contrastive learning pays off when the encoder involves a complex
+architecture, like RestNet, and multi-class problems with many labels.
```
Can you please change `RestNet` to `ResNet` (assuming you're discussing residual networks like in https://arxiv.org/abs/1512.03385)?

```suggestion
architecture, like ResNet, and multi-class problems with many labels.
```

ksalama replied:
Thanks @8bitmp3 for spotting this. I have applied the fixes 👍

8bitmp3 reviewed (changes requested):
Hey @ksalama I'm a tf.keras user here and just came across the PR. This example is awesome. 

Can you please check and change `RestNet` to `ResNet` (assuming you're discussing residual networks like in https://arxiv.org/abs/1512.03385).

```
architecture, like ResNet, and multi-class problems with many labels.
```

Can I also suggest to improve the introduction just a little bit to help the readers understand what they're about to learn? You could, for example, write about 1) what the user will learn about - in a bit more detail (can be summarized from the paper on arXiv (with the link included) before going into the two-phase structure that you already mentioned; 2) what the user will do - it's a classification task for X using Y dataset. Both of these could be just 1-2 short sentences.

Also, I'd advise to copy-paste the Markdown text into Google Docs just to check that all grammar and spelling is 👍 - but I've not spotted anything out of the ordinary here.

Anyway, these are just suggestions from experience. I really enjoyed the paper, thanks for sharing the work!

fchollet commented:
Thanks @8bitmp3 for the feedback. @ksalama, I believe the example is ready, so you can add the `.md` and `.ipynb` files to the PR.

ksalama commented:
@fchollet I have added the generated .md and .ipynb files to the PR

fchollet reviewed (approved):
Excellent, thank you for the great contribution! I'll take it from there 👍
